### PR TITLE
K8SPSMDB-198 Version update

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-198](https://jira.percona.com/browse/K8SPSMDB-198)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | failed |
| self-healing | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| scaling | failed |
| security-context | passed |
| arbiter | passed |
| operator-self-healing | passed |
| scheduled-backup | failed |
| service-per-pod | failed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| monitoring | passed |
| init-deploy | passed |
| limits | passed |
| scaling | failed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade | failed |
| security-context | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| limits | passed |
| self-healing | passed |
| scaling | failed |
| operator-self-healing | passed |
| upgrade | failed |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| security-context | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | failed |
| one-pod | passed |
| upgrade | failed |
| self-healing | failed |
| limits | passed |
| operator-self-healing | passed |
| monitoring | failed |
| scaling | failed |
| security-context | failed |
| monitoring-2-0 | failed |
| arbiter | failed |
| service-per-pod | failed |
| liveness | passed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| storage | passed |
| one-pod | failed |
| upgrade | failed |
| limits | passed |
| self-healing | failed |
| upgrade-consistency | passed |
| scaling | failed |
| demand-backup | failed |
| scheduled-backup | failed |
| operator-self-healing | passed |
| security-context | passed |
| monitoring | failed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| storage | failed |
| one-pod | failed |
| upgrade | failed |
| monitoring | failed |
| limits | passed |
| self-healing | passed |
| scaling | passed |
| monitoring-2-0 | failed |
| security-context | passed |
| arbiter | passed |
| operator-self-healing | passed |
| service-per-pod | failed |
| liveness | failed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| self-healing | passed |
| limits | failed |
| operator-self-healing | passed |
| scaling | passed |
| security-context | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| upgrade | failed |
| storage | passed |
| monitoring | failed |
| self-healing | passed |
| limits | passed |
| monitoring-2-0 | failed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | failed |
| liveness | failed |
| upgrade-consistency | failed |

File: pkg/controller/perconaservermongodb/psmdb_controller.go [hunk 1]
```diff
@@ -193,7 +220,7 @@ func (r *ReconcilePerconaServerMongoDB) Reconcile(request reconcile.Request) (re
 			return reconcile.Result{}, err
 		}
 	}
-
+	var sfs *appsv1.StatefulSet
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/perconaservermongodb/psmdb_controller.go]: https://github.com/percona/percona-server-mongodb-operator/pull/360/commits/ddc997cd507a1969a9644b0d0ac233b72b25b5c3

File: pkg/controller/perconaservermongodb/version.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+		err := r.client.Get(context.TODO(), types.NamespacedName{Name: cr.Name, Namespace: cr.Namespace}, localCr)
+		if err != nil {
+			log.Error(err, "failed to get CR")
+			return
+		}
+
+		if localCr.Status.MongoStatus != v1.AppStateReady {
+			log.Info("cluster is not ready")
+			return
+		}
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/perconaservermongodb/version.go]: https://github.com/percona/percona-server-mongodb-operator/pull/360/commits/7ba8b58472f240d6adcd444e6999e62df7a0f240

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | failed |
| init-deploy | failed |
| upgrade | failed |
| limits | passed |
| monitoring | failed |
| self-healing | passed |
| scaling | failed |
| upgrade-consistency | passed |
| security-context | failed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| operator-self-healing | failed |
| arbiter | failed |
| service-per-pod | failed |
| scheduled-backup | failed |
| liveness | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| self-healing | passed |
| limits | passed |
| monitoring | failed |
| scaling | passed |
| monitoring-2-0 | failed |
| upgrade | failed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | failed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| init-deploy | failed |
| storage | passed |
| upgrade | failed |
| limits | failed |
| monitoring | failed |
| scaling | failed |
| self-healing | failed |
| upgrade-consistency | failed |
| demand-backup | failed |
| security-context | passed |
| scheduled-backup | failed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | failed |
| one-pod | failed |
| upgrade | failed |
| init-deploy | failed |
| monitoring | failed |
| operator-self-healing | passed |
| monitoring-2-0 | failed |
| limits | passed |
| scaling | failed |
| arbiter | passed |
| service-per-pod | failed |
| security-context | passed |
| liveness | passed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade | failed |
| monitoring | passed |
| limits | passed |
| scaling | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | passed |
| upgrade-consistency | failed |

File: pkg/controller/perconaservermongodb/version.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+		cr.Spec.UpgradeOptions.Apply == never ||
+		cr.Spec.UpgradeOptions.Apply == disabled {
+		return nil
+	}
+
+	shedule, ok := r.crons.jobs[jobName]
+	if ok && (cr.Spec.UpgradeOptions.Schedule == "" ||
+		cr.Spec.UpgradeOptions.Apply == never ||
+		cr.Spec.UpgradeOptions.Apply == disabled) {
+		r.deleteEnsureVersion(shedule.ID)
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/perconaservermongodb/version.go]: https://github.com/percona/percona-server-mongodb-operator/pull/360/commits/55cd6d867d3723e845723e77e4bd5a6e0cb074ac

File: pkg/controller/perconaservermongodb/version.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+func (r *ReconcilePerconaServerMongoDB) sheduleEnsureVersion(cr *api.PerconaServerMongoDB, vs VersionService) error {
+	if cr.Spec.UpdateStrategy != api.SmartUpdateStatefulSetStrategyType ||
+		cr.Spec.UpgradeOptions.Schedule == "" ||
+		cr.Spec.UpgradeOptions.Apply == never ||
+		cr.Spec.UpgradeOptions.Apply == disabled {
+		return nil
+	}
+
+	shedule, ok := r.crons.jobs[jobName]
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/perconaservermongodb/version.go]: fixed

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | failed |
| init-deploy | failed |
| upgrade | failed |
| one-pod | passed |
| limits | passed |
| monitoring | failed |
| self-healing | passed |
| scaling | failed |
| monitoring-2-0 | failed |
| security-context | failed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| upgrade-consistency | failed |

File: pkg/controller/perconaservermongodb/version.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	api "github.com/percona/percona-server-mongodb-operator/pkg/apis/psmdb/v1"
+	v1 "github.com/percona/percona-server-mongodb-operator/pkg/apis/psmdb/v1"
+	"github.com/pkg/errors"
+	"github.com/robfig/cron/v3"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const jobName = "ensure-version"
+const never = "Never"
+const disabled = "Disabled"
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/perconaservermongodb/version.go]: https://github.com/percona/percona-server-mongodb-operator/pull/360/commits/cdfe0ca6ffb5556812c35e9de6fe4a5cf400ea27

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | failed |
| self-healing | failed |
| monitoring | failed |
| limits | passed |
| monitoring-2-0 | failed |
| scaling | passed |
| operator-self-healing | passed |
| arbiter | passed |
| security-context | passed |
| service-per-pod | passed |
| liveness | passed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| monitoring | passed |
| upgrade | failed |
| limits | passed |
| monitoring-2-0 | failed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| service-per-pod | passed |
| liveness | passed |
| upgrade-consistency | failed |

**Pull request merged**